release.py

```diff
@@ -0,0 +1,241 @@
+            )
+            print(f"{bcolors.FAIL} {e} {bcolors.ENDC}")
+    subprocess.run(["git", "checkout", BRANCH])
+    if args.retry and os.path.isfile(RELEASE_PROGRESS + "/tox_news_complete"):
+        print(f"{bcolors.OKCYAN}Skipping news generation...{bcolors.ENDC}")
+    else:
+        try:
+            subprocess.run(["./venv/bin/tox", "-e", "news"], check=True)
+        except Exception as e:
+            print(f"{bcolors.FAIL}INFO: Failed to generate news! :(...{bcolors.ENDC}")
+            print(f"{bcolors.FAIL} {e} {bcolors.ENDC}")
+    if args.retry and os.path.isfile(RELEASE_PROGRESS + "/newsfragment_complete"):
+        print(f"{bcolors.OKCYAN}Skipping add news fragment...{bcolors.ENDC}")
+    else:
+        try:
+            subprocess.run(
+                ["touch", "newsfragments/{ticket}.minor".format(ticket=TICKET)],
+                check=True,
+            )
+        except Exception as e:
+            print(
+                f"{bcolors.FAIL}INFO: Failed to add newsfragment file! :(...{bcolors.ENDC}"
+            )
+            print(f"{bcolors.FAIL} {e} {bcolors.ENDC}")
+    if args.retry and os.path.isfile(
+        RELEASE_PROGRESS + "/commit_newsfragment_complete"
+    ):
+        print(f"{bcolors.OKCYAN}Skipping commit fragment...{bcolors.ENDC}")
+    else:
+        try:
+            subprocess.run(["git", "add", "."], check=True)
+            subprocess.run(
+                ["git", "commit", "-s", "-m", "tahoe-lafs-{tag} news".format(tag=TAG)],
+                check=True,
+            )
+        except Exception as e:
+            print(
+                f"{bcolors.FAIL}INFO: Failed to commit newsfragment! :(...{bcolors.ENDC}"
+            )
+            print(f"{bcolors.FAIL} {e} {bcolors.ENDC}")
+    updated_content = None
+    LINE = ""
+    for i in range(0, len(RELEASE_TITLE)):
+        LINE += "="
+        i += 1
+    LINE = "\n" + LINE
+    with open("NEWS.rst", "r") as f:
+        content = f.read()
+        updated_content = re.sub(
+            r"\.\.\stowncrier start line\n(Release\s(\d+\.\d+\.\d)(\.)post\d+\s\(\d{4}-\d{02}-\d{02}\))+(\n\'+)",
+            RELEASE_TITLE + LINE,
+            content,
+        )
+    with open("updated_news.rst", "w") as f:
+        f.write(updated_content)
+    subprocess.run(["mv", "updated_news.rst", "NEWS.rst"])
+    print(f"{bcolors.OKGREEN}First release step complete.{bcolors.ENDC}")
+    print(f"{bcolors.OKBLUE}Instruction: Please review News.rst{bcolors.ENDC}")
+    print(
+        f'{bcolors.OKBLUE}Instruction: Update "docs/known_issues.rst" (if neccesary){bcolors.ENDC}'
+    )
+    print(
+        f'{bcolors.OKBLUE}Instruction: If any, commit changes to NEWS.rst and  "docs/known_issues.rst" {bcolors.ENDC}'
+    )
+    print(
+        f"{bcolors.OKBLUE}Instruction: Make sure your github username, password and gpg passphrase are ready for the next step."
+    )
+    print(CONTINUE_INSTRUCTION)
+
+
+def complete_release():
+    os.chdir(RELEASE_FOLDER)
```

Comment:
Should put the directory back (e.g. when done in here) probably with a try/finally?

Reply:
Not sure what you mean here.

Reply:
sorry, was thinking of `pushd` / `popd` .. the shell itself should be in the same directory on exit